File: interactive_solver.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
حل‌کننده تعاملی برای مسئله مسیرهای منحصر به فرد با موانع
"""

def get_matrix_from_user():
    """
    دریافت ماتریس از کاربر
    """
    print("لطفاً ابعاد ماتریس را وارد کنید:")
    
    try:
        m = int(input("تعداد ردیف‌ها (m): "))
        n = int(input("تعداد ستون‌ها (n): "))
        
        if m <= 0 or n <= 0:
            print("ابعاد باید مثبت باشند!")
            return None
        
        print(f"\nماتریس {m}×{n} را وارد کنید:")
        print("0 = مسیر آزاد، 1 = مانع")
        print("هر ردیف را در یک خط با فاصله وارد کنید:")
        
        matrix = []
        for i in range(m):
            while True:
                try:
                    row_input = input(f"ردیف {i+1}: ")
                    row = list(map(int, row_input.split()))
                    
                    if len(row) != n:
                        print(f"باید دقیقاً {n} عدد وارد کنید!")
                        continue
                    
                    if not all(x in [0, 1] for x in row):
                        print("فقط اعداد 0 و 1 مجاز هستند!")
                        continue
                    
                    matrix.append(row)
                    break
                except ValueError:
                    print("لطفاً فقط اعداد صحیح وارد کنید!")
        
        return matrix
        
    except ValueError:
        print("لطفاً عدد صحیح وارد کنید!")
        return None


def unique_paths_with_obstacles(obstacle_grid):
    """
    محاسبه مسیرهای منحصر به فرد
    """
    if not obstacle_grid or not obstacle_grid[0]:
        return 0
    
    m, n = len(obstacle_grid), len(obstacle_grid[0])
    
    if obstacle_grid[0][0] == 1 or obstacle_grid[m-1][n-1] == 1:
        return 0
    
    dp = [[0] * n for _ in range(m)]
    dp[0][0] = 1
    
    # ردیف اول
    for j in range(1, n):
        if obstacle_grid[0][j] == 0:
            dp[0][j] = dp[0][j-1]
    
    # ستون اول
    for i in range(1, m):
        if obstacle_grid[i][0] == 0:
            dp[i][0] = dp[i-1][0]
    
    # باقی جدول
    for i in range(1, m):
        for j in range(1, n):
            if obstacle_grid[i][j] == 0:
                dp[i][j] = dp[i-1][j] + dp[i][j-1]
    
    return dp[m-1][n-1], dp


def print_detailed_solution(obstacle_grid, dp_matrix):
    """
    نمایش جزئیات حل
    """
    m, n = len(obstacle_grid), len(obstacle_grid[0])
    
    print("\n" + "="*50)
    print("نتایج تفصیلی:")
    print("="*50)
    
    print("\n1. ماتریس ورودی:")
    for i in range(m):
        print("   ", end="")
        for j in range(n):
            if obstacle_grid[i][j] == 1:
                print("🚫", end=" ")
            else:
                print("⬜", end=" ")
        print()
    
    print("\n2. جدول مسیرها (DP):")
    for i in range(m):
        print("   ", end="")
        for j in range(n):
            if obstacle_grid[i][j] == 1:
                print("  X", end=" ")
            else:
                print(f"{dp_matrix[i][j]:3d}", end=" ")
        print()
    
    print(f"\n3. تعداد کل مسیرها: {dp_matrix[m-1][n-1]}")


def main():
    """
    برنامه اصلی
    """
    print("🤖 حل‌کننده مسئله مسیرهای منحصر به فرد با موانع")
    print("="*60)
    
    while True:
        print("\nگزینه‌ها:")
        print("1. ورود ماتریس جدید")
        print("2. استفاده از مثال آماده")
        print("3. خروج")
        
        choice = input("\nانتخاب کنید (1-3): ")
        
        if choice == "1":
            matrix = get_matrix_from_user()
            if matrix:
                result, dp_matrix = unique_paths_with_obstacles(matrix)
                print_detailed_solution(matrix, dp_matrix)
        
        elif choice == "2":
            examples = [
                {
                    "name": "مثال ساده",
                    "grid": [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
                },
                {
                    "name": "مثال پیچیده",
                    "grid": [[0, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 0]]
                },
                {
                    "name": "مسیر مسدود",
                    "grid": [[0, 1], [0, 0]]
                }
            ]
            
            print("\nمثال‌های آماده:")
            for i, example in enumerate(examples, 1):
                print(f"{i}. {example['name']}")
            
            try:
                example_choice = int(input("انتخاب کنید: ")) - 1
                if 0 <= example_choice < len(examples):
                    selected = examples[example_choice]
                    print(f"\n{selected['name']} انتخاب شد:")
                    result, dp_matrix = unique_paths_with_obstacles(selected['grid'])
                    print_detailed_solution(selected['grid'], dp_matrix)
                else:
                    print("انتخاب نامعتبر!")
            except ValueError:
                print("لطفاً عدد صحیح وارد کنید!")
        
        elif choice == "3":
            print("خداحافظ! 👋")
            break
        
        else:
            print("انتخاب نامعتبر! لطفاً 1، 2 یا 3 را انتخاب کنید.")


if __name__ == "__main__":
    main()

File: InteractiveSolver.java
import java.util.Scanner;

/**
 * حل‌کننده تعاملی برای مسئله مسیرهای منحصر به فرد با موانع
 */
public class InteractiveSolver {
    
    private static Scanner scanner = new Scanner(System.in);
    
    /**
     * دریافت ماتریس از کاربر
     */
    public static int[][] getMatrixFromUser() {
        System.out.println("لطفاً ابعاد ماتریس را وارد کنید:");
        
        try {
            System.out.print("تعداد ردیف‌ها (m): ");
            int m = scanner.nextInt();
            System.out.print("تعداد ستون‌ها (n): ");
            int n = scanner.nextInt();
            
            if (m <= 0 || n <= 0) {
                System.out.println("ابعاد باید مثبت باشند!");
                return null;
            }
            
            System.out.println("\nماتریس " + m + "×" + n + " را وارد کنید:");
            System.out.println("0 = مسیر آزاد، 1 = مانع");
            System.out.println("هر ردیف را در یک خط با فاصله وارد کنید:");
            
            int[][] matrix = new int[m][n];
            
            for (int i = 0; i < m; i++) {
                System.out.print("ردیف " + (i+1) + ": ");
                for (int j = 0; j < n; j++) {
                    int value = scanner.nextInt();
                    if (value != 0 && value != 1) {
                        System.out.println("فقط اعداد 0 و 1 مجاز هستند!");
                        return null;
                    }
                    matrix[i][j] = value;
                }
            }
            
            return matrix;
            
        } catch (Exception e) {
            System.out.println("خطا در ورودی! لطفاً دوباره تلاش کنید.");
            scanner.nextLine(); // پاک کردن بافر
            return null;
        }
    }
    
    /**
     * محاسبه مسیرهای منحصر به فرد
     */
    public static int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid == null || obstacleGrid.length == 0 || obstacleGrid[0].length == 0) {
            return 0;
        }
        
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        
        if (obstacleGrid[0][0] == 1 || obstacleGrid[m-1][n-1] == 1) {
            return 0;
        }
        
        int[][] dp = new int[m][n];
        dp[0][0] = 1;
        
        // ردیف اول
        for (int j = 1; j < n; j++) {
            if (obstacleGrid[0][j] == 0) {
                dp[0][j] = dp[0][j-1];
            }
        }
        
        // ستون اول
        for (int i = 1; i < m; i++) {
            if (obstacleGrid[i][0] == 0) {
                dp[i][0] = dp[i-1][0];
            }
        }
        
        // باقی جدول
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (obstacleGrid[i][j] == 0) {
                    dp[i][j] = dp[i-1][j] + dp[i][j-1];
                }
            }
        }
        
        return dp[m-1][n-1];
    }
    
    /**
     * نمایش جزئیات حل
     */
    public static void printDetailedSolution(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        
        System.out.println("\n" + "=".repeat(50));
        System.out.println("نتایج تفصیلی:");
        System.out.println("=".repeat(50));
        
        System.out.println("\n1. ماتریس ورودی:");
        for (int i = 0; i < m; i++) {
            System.out.print("   ");
            for (int j = 0; j < n; j++) {
                if (obstacleGrid[i][j] == 1) {
                    System.out.print("🚫 ");
                } else {
                    System.out.print("⬜ ");
                }
            }
            System.out.println();
        }
        
        // محاسبه جدول DP
        if (obstacleGrid[0][0] == 1 || obstacleGrid[m-1][n-1] == 1) {
            System.out.println("\nنقطه شروع یا پایان مسدود است!");
            return;
        }
        
        int[][] dp = new int[m][n];
        dp[0][0] = 1;
        
        // پر کردن جدول DP
        for (int j = 1; j < n; j++) {
            if (obstacleGrid[0][j] == 0) {
                dp[0][j] = dp[0][j-1];
            }
        }
        
        for (int i = 1; i < m; i++) {
            if (obstacleGrid[i][0] == 0) {
                dp[i][0] = dp[i-1][0];
            }
        }
        
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                if (obstacleGrid[i][j] == 0) {
                    dp[i][j] = dp[i-1][j] + dp[i][j-1];
                }
            }
        }
        
        System.out.println("\n2. جدول مسیرها (DP):");
        for (int i = 0; i < m; i++) {
            System.out.print("   ");
            for (int j = 0; j < n; j++) {
                if (obstacleGrid[i][j] == 1) {
                    System.out.print("  X ");
                } else {
                    System.out.printf("%3d ", dp[i][j]);
                }
            }
            System.out.println();
        }
        
        System.out.println("\n3. تعداد کل مسیرها: " + dp[m-1][n-1]);
    }
    
    /**
     * اجرای مثال‌های آماده
     */
    public static void runPredefinedExamples() {
        int[][][] examples = {
            {{0, 0, 0}, {0, 1, 0}, {0, 0, 0}},  // مثال ساده
            {{0, 0, 0, 0}, {0, 1, 0, 0}, {0, 0, 1, 0}, {0, 0, 0, 0}},  // مثال پیچیده
            {{0, 1}, {0, 0}},  // مسیر مسدود
            {{1, 0}, {0, 0}}   // شروع مسدود
        };
        
        String[] names = {
            "مثال ساده (3×3)",
            "مثال پیچیده (4×4)", 
            "مسیر مسدود (2×2)",
            "شروع مسدود (2×2)"
        };
        
        System.out.println("\nمثال‌های آماده:");
        for (int i = 0; i < examples.length; i++) {
            System.out.println((i+1) + ". " + names[i]);
        }
        
        System.out.print("انتخاب کنید (1-" + examples.length + "): ");
        try {
            int choice = scanner.nextInt() - 1;
            if (choice >= 0 && choice < examples.length) {
                System.out.println("\n" + names[choice] + " انتخاب شد:");
                printDetailedSolution(examples[choice]);
            } else {
                System.out.println("انتخاب نامعتبر!");
            }
        } catch (Exception e) {
            System.out.println("ورودی نامعتبر!");
            scanner.nextLine();
        }
    }
    
    /**
     * متد اصلی
     */
    public static void main(String[] args) {
        System.out.println("🤖 حل‌کننده مسئله مسیرهای منحصر به فرد با موانع");
        System.out.println("=".repeat(60));
        
        while (true) {
            System.out.println("\nگزینه‌ها:");
            System.out.println("1. ورود ماتریس جدید");
            System.out.println("2. استفاده از مثال آماده");
            System.out.println("3. خروج");
            
            System.out.print("\nانتخاب کنید (1-3): ");
            
            try {
                int choice = scanner.nextInt();
                
                switch (choice) {
                    case 1:
                        int[][] matrix = getMatrixFromUser();
                        if (matrix != null) {
                            printDetailedSolution(matrix);
                        }
                        break;
                    
                    case 2:
                        runPredefinedExamples();
                        break;
                    
                    case 3:
                        System.out.println("خداحافظ! 👋");
                        scanner.close();
                        return;
                    
                    default:
                        System.out.println("انتخاب نامعتبر! لطفاً 1، 2 یا 3 را انتخاب کنید.");
                }
            } catch (Exception e) {
                System.out.println("ورودی نامعتبر! لطفاً عدد صحیح وارد کنید.");
                scanner.nextLine(); // پاک کردن بافر
            }
        }
    }
}